File: String List Editor/string_list_editor.java
import java.util.Arrays;

public class MIM_Code {
	
	public static void main(String[] args)
	{
		String[] strings = {"HelloThere", "name", "YELling","hyphen-","ate"};	//sample input with all different cases
		string_list_editor(strings);
	}
	
	public static void string_list_editor(String[] strings)
	{
		int start_chars = 0, end_chars = 0, median = 0;
		start_chars = count_chars(strings);		
		
		int[] lengths = length_array(strings);	
		if(lengths.length % 2 == 0)			//calculate median
			median = (lengths[lengths.length / 2] + lengths[(lengths.length/2)-1]) / 2;
		else
			median = lengths[lengths.length / 2];
		
		if (strings.length > 0)			//check for valid inputs
		{
			for (int i = 0; i < strings.length; i++)
			{
				if(strings[i].length() % 4 == 0)	//length divisible by 4
				{
					String new_string = "";
					for (int j = strings[i].length() - 1; j >= 0; j--)
					{
						new_string += strings[i].charAt(j);
					}
					strings[i] = new_string;
				}
//*****************************************************************************************				
	
				if(strings[i].length() % 5 == 0)		//length divisible by 5
				{
					strings[i] = strings[i].substring(0,5);
				}
//*****************************************************************************************	
				
				int count = 0;			//check for capital letters
				for(int k = 0; k < strings[i].length(); k++)
				{
					if (k == 5)
					{
						break;
					}
					if (Character.isUpperCase(strings[i].charAt(k)))
					{
						count++;
					}
				}
				if (count >= 3)		//if 3 capitals, capitalize word
				{
					strings[i] = strings[i].toUpperCase();
				}
//*******************************************************************************************	
				
				if(strings[i].charAt(strings[i].length() - 1) == '-') //check for ending hyphen
				{
					if(i != strings.length - 1)
					{
						strings[i] = strings[i].substring(0,strings[i].length() - 1) + strings[i+1];
					}
					else
					{
						strings[i] = strings[i].substring(0,strings[i].length() - 1);
					}
				}
//*******************************************************************************************
				
				System.out.println(strings[i]);
			}
			end_chars = count_chars(strings);
			System.out.println("Starting chars: " + start_chars + "\nEnding chars: " + end_chars + "\nMedian: " + median);
		}
	}
	
	public static int count_chars(String[] words)		//counts chars in list
	{
		int num = 0;
		for (int index = 0; index < words.length; index++)
		{
			num += words[index].length();
		}
		return num;
	}
	
	public static int[] length_array(String[] words)	//returns a list of the lengths
	{
		int[] nums = new int[words.length];
		for(int i = 0; i < words.length; i++)
		{
			nums[i] = words[i].length();
		}
		Arrays.sort(nums);
		return nums;
	}
}

File: String List Editor/string_list_editor.py
def string_list_editor(strings):
	start_chars = count_chars(strings)	
	arranged = []

	for string in strings:		#sort the input list to find median
		arranged.append(len(string))
	sorted(arranged)

	if len(arranged) % 2 == 0:			#calculate median
		median = (arranged[len(arranged)/2] + arranged[(len(arranged)/2) - 1]) / 2
	else:
		median = arranged[len(arranged)//2]

	if len(strings) > 0:				##check for bad inputs
		for index in range(len(strings)):
			if len(strings[index]) % 4 == 0:			##string length divisible by 4
				strings[index] = strings[index][::-1]
##****************************************************************************************

			if len(strings[index]) % 5 == 0:			##string length divisible by 5
				strings[index] = strings[index][:5]
##******************************************************************************************

			count = 0
			for char in strings[index][:5]:				##count number of capitals in first 5
				if char == char.upper():
					count += 1
			if count >= 3:								##make word a capital if true
				strings[index] = strings[index].upper()
				count = 0
##**********************************************************************************************

			if strings[index][len(strings[index])-1:] == "-":	##check for hyphen
				if index == len(strings) - 1:					##if there is no next word, just remove hyphen
					strings[index] = strings[index][:len(strings)-1]
				else:
					strings[index] = strings[index][:len(strings[index])-1]
					strings[index] += strings[index + 1]
##**********************************************************************************************

			print (strings[index])

	end_chars = count_chars(strings)
	print("Starting chars: " + str(start_chars) + "\nEnding Chars: " + str(end_chars) + "\nMedian length of input: " + str(median))

def count_chars(strings):		##counts chars in a list
	total = 0
	for string in strings:
		total += len(string)
	return total

string_list_editor(["hellothere","name","YELling","hyphen-","ate"])		##sample input with one case for each objective